.github/workflows/pypi_upload.yml

```diff
@@ -0,0 +1,78 @@
+name: upload-pypi
+
+on:
+  workflow_call:
+    inputs:
+      repository:
+        description: Repository to checkout
+        default: ''
+        type: string
+      ref:
+        description: Reference to checkout
+        default: nightly
+        type: string
+      test-infra-repository:
+        description: Test infra repository to use
+        default: pytorch/test-infra
+        type: string
+      test-infra-ref:
+        description: Test infra reference to use
+        default: ''
+        type: string
+      build-matrix:
+        description: Build matrix to utilize
+        default: ''
+        type: string
+      architecture:
+        description: Architecture to build for x86_64 for default Linux, or aarch64 for Linux aarch64 builds
+        required: false
+        type: string
+        default: x86_64
+      trigger-event:
+        description: Trigger event in caller that determines whether or not to upload
+        type: string
+        default: ''
+
+jobs:
+  upload:
+    runs-on: ubuntu-22.04
+    strategy:
+      fail-fast: false
+      matrix: ${{ fromJSON(inputs.build-matrix) }}
+    timeout-minutes: 30
+    name: ${{ matrix.build_name }}
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          repository: ${{ inputs.test-infra-repository }}
+          ref: ${{ inputs.test-infra-ref }}
+          path: test-infra
+
+      - uses: ./test-infra/.github/actions/set-channel
+
+      - uses: ./test-infra/.github/actions/setup-binary-upload
+        with:
+          repository: ${{ inputs.repository }}
+          ref: ${{ inputs.ref }}
+          python-version: ${{ matrix.python_version }}
+          cuda-version: ${{ matrix.desired_cuda }}
+          arch: ${{ inputs.architecture }}
+
+      - name: Download the artifact
+        uses: actions/download-artifact@v3
+        with:
+          name: ${{ env.ARTIFACT_NAME }}
+          path: ${{ inputs.repository }}/dist/
+
+      - name: Upload CUDA 12.1 package to pypi on schedule
+        # TODO (huydhn): This is a work around to upload only CUDA 12.1 to pypi.
+        # Other binaries for different CUDA versions will be available on
+        # download.pytorch.org
+        if: ${{ inputs.trigger-event == 'schedule' && matrix.desired_cuda == 'cu121' }}
```

Comment:
Wait so does this mean that we spawn a bunch of jobs here to upload to pypi but only actually do the upload for `cu121`?

Why not just make it so it only does the upload for the specific one we want instead of spawning a bunch of jobs that are basically no-ops?

Reply:
Well, I tried to skip those no-op upload, but thought that checking for `cu121` was like a hack anyway.  So, I plan to make change further upstream to `pytorch/test-infra/.github/workflows/generate_binary_build_matrix.yml@main` later to limit the cuda version there instead.

Anyway, I think those redundant jobs will be dealt with after I update test-infra upload step to handle pypi upload.  A small refactor is on the way.

Reply:
So I do want us to mimic what torch does. So for example in our official releases use cuda 11.8 and 12.1. Not sure if policy is different for cuda versions between nightlies and official release

Reply:
In PyTorch, both nightly and official releases can be installed from https://download.pytorch.org for all CUDA versions.  We then upload to pypi only one blessed CUDA version, which is 12.1 atm https://pypi.org/project/torch/2.3.0/#files.  So, what we have here indeed mimic what PyTorch has.